File: blind75/41_remove_nth_node_from_end_of_list/remove_nth_node.go
package remove_nth_node

// ListNode represents a node in a singly linked list.
type ListNode struct {
	Val  int
	Next *ListNode
}

// RemoveNthFromEnd removes the nth node from the end using two pointers (one pass).
// Time complexity: O(n), Space complexity: O(1)
func RemoveNthFromEnd(head *ListNode, n int) *ListNode {
	dummy := &ListNode{Next: head}
	fast := dummy
	slow := dummy

	// Move fast n+1 steps ahead
	for i := 0; i <= n; i++ {
		fast = fast.Next
	}

	// Move both pointers until fast reaches the end
	for fast != nil {
		fast = fast.Next
		slow = slow.Next
	}

	// Remove the nth node
	slow.Next = slow.Next.Next

	return dummy.Next
}

// RemoveNthFromEndTwoPass removes the nth node using two passes.
// Time complexity: O(n), Space complexity: O(1)
func RemoveNthFromEndTwoPass(head *ListNode, n int) *ListNode {
	// First pass: count the length
	length := 0
	curr := head
	for curr != nil {
		length++
		curr = curr.Next
	}

	// Edge case: remove the first node
	if n == length {
		return head.Next
	}

	// Second pass: move to the node before the one to remove
	dummy := &ListNode{Next: head}
	curr = dummy
	for i := 0; i < length-n; i++ {
		curr = curr.Next
	}

	// Remove the node
	curr.Next = curr.Next.Next

	return dummy.Next
}

File: blind75/41_remove_nth_node_from_end_of_list/remove_nth_node.py
from typing import Optional


class ListNode:
    """Definition for singly-linked list node."""

    def __init__(self, val: int = 0, next: Optional['ListNode'] = None):
        self.val = val
        self.next = next


def remove_nth_from_end(head: Optional[ListNode], n: int) -> Optional[ListNode]:
    """
    Removes the nth node from the end using two pointers (one pass).
    Time complexity: O(n), Space complexity: O(1)

    Args:
        head: The head of the linked list
        n: The position from the end (1-indexed)

    Returns:
        The head of the modified linked list
    """
    dummy = ListNode(0, head)
    fast = dummy
    slow = dummy

    # Move fast n+1 steps ahead
    for _ in range(n + 1):
        fast = fast.next

    # Move both pointers until fast reaches the end
    while fast is not None:
        fast = fast.next
        slow = slow.next

    # Remove the nth node
    slow.next = slow.next.next

    return dummy.next


def remove_nth_from_end_two_pass(head: Optional[ListNode], n: int) -> Optional[ListNode]:
    """
    Removes the nth node from the end using two passes.
    Time complexity: O(n), Space complexity: O(1)

    Args:
        head: The head of the linked list
        n: The position from the end (1-indexed)

    Returns:
        The head of the modified linked list
    """
    # First pass: count the length
    length = 0
    curr = head
    while curr is not None:
        length += 1
        curr = curr.next

    # Edge case: remove the first node
    if n == length:
        return head.next

    # Second pass: move to the node before the one to remove
    dummy = ListNode(0, head)
    curr = dummy
    for _ in range(length - n):
        curr = curr.next

    # Remove the node
    curr.next = curr.next.next

    return dummy.next
